### PR TITLE
ARROW-15550: [C++] Add optional debug memory checks

### PR DESCRIPTION
**Pull request by pitrou**
Debug memory checks can be enabled by setting the environment variable ARROW_DEBUG_MEMORY_POOL
to one of the recognized values "abort", "trap" or "warn" (regardless of whether Arrow
was compiled in debug or release mode).

The only implemented check adds a suffix past the allocation and checks that it isn't clobbered on deallocation.
Adding a prefix would be more expensive because of alignment constraints.

Also add a fix by Will Jones for ARROW-14047 (Parquet Arrow reader sets null values in buffer overflow), since otherwise CI would fail.

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Review by wjones127 — APPROVED** [review]
This looks great! GDB integration sounds cool (though I haven't tried it out yet).

File: cpp/src/parquet/level_conversion_inc.h [hunk 1]
```diff
@@ -333,7 +333,7 @@ void DefLevelsToBitmapSimd(const int16_t* def_levels, int64_t num_def_levels,
   ::arrow::internal::FirstTimeBitmapWriter writer(
       output->valid_bits,
       /*start_offset=*/output->valid_bits_offset,
-      /*length=*/num_def_levels);
+      /*length=*/output->values_read_upper_bound);
```
[the remark this answers is not in the document]
Comment (lidavidm) [reply cpp/src/parquet/level_conversion_inc.h:336]: Ah - it hasn't yet been fixed on master - and presumably it fails without this fix.

File: python/pyarrow/tests/test_memory.py [hunk 2]
```diff
@@ -169,3 +178,63 @@ def test_supported_memory_backends():
...
     if should_have_mimalloc:
         assert "mimalloc" in backends
+
+
+def run_debug_memory_pool(pool_factory, env_value):
+    """
+    Run a piece of code making an invalid memory write with the
+    ARROW_DEBUG_MEMORY_POOL environment variable set to a specific value.
+    """
+    code = f"""if 1:
```
Comment (lidavidm) [inline python/pyarrow/tests/test_memory.py:188]: is the `if 1` just so the string contents don't have to be dedented?

**Review by lidavidm — APPROVED** [review]

File: python/pyarrow/tests/test_memory.py (hunk 2 above)
Comment (pitrou) [reply python/pyarrow/tests/test_memory.py:188]: Yes!

**Comment by pitrou** [comment]
Yes, the conda-cpp-valgrind failure is https://issues.apache.org/jira/browse/ARROW-15354

**Pull request closed without merging**